Gemini cli not detecting IntelliJ on linux

I'm continuing [the conversation from the `gemini-cli` MR](https://github.com/google-gemini/gemini-cli/issues/9273#issuecomment-3666603286).

I can help debugging, but if you give me some pointers on how you debug it, it will be easier.

Thanks.

> Running gemini from the terminal didn't change the result, previous time when I started it from the button I already had a couple of terminals open.
> 
> I'm using IntelliJ.let me know if I can help debug and the steps I should take for that.
> 
> Thanks.
>

I tried opening the project in IntelliJ to debug, but I can't debug it, when I try to refresh the project with gradle I just get this:

```
Cause: error=2, No existe el archivo o el directorio
```

I can't see what directory or file does not exist.

And when I manually try to generate the idea project using gradle:
```
$ gradle idea     
executing gradlew instead of gradle
Starting a Gradle Daemon, 4 stopped Daemons could not be reused, use --status for details
Calculating task graph as no cached configuration is available for tasks: idea

> Task :idea
Generated IDEA project at file:///media/disk/Documents/personal-projects/jetbrains-ide-companion/gemini-cli-jetbrains-companion.ipr

[Incubating] Problems report is available at: file:///media/disk/Documents/personal-projects/jetbrains-ide-companion/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache problems found in this build.

6 problems were found storing the configuration cache, 3 of which seem unique.
- Task `:ideaModule` of type `org.gradle.plugins.ide.idea.GenerateIdeaModule`: cannot serialize object of type 'org.gradle.api.internal.project.DefaultProject', a subtype of 'org.gradle.api.Project', as these are not supported with the configuration cache.
  See https://docs.gradle.org/8.14.3/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
- Task `:ideaProject` of type `org.gradle.plugins.ide.idea.GenerateIdeaProject`: Class 'org.gradle.plugins.ide.idea.IdeaPlugin$$Lambda/0x0000786108ad9a28' cannot be encoded because Gradle runtime class loader 'MixInLegacyTypesClassLoader(legacy-mixin-loader)' of type 'org.gradle.initialization.MixInLegacyTypesClassLoader' could not be encoded and the class is not available through the default class loader.
These are the known class loaders:
        - VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:]:root-project[:]:kotlin-dsl:/media/disk/Documents/personal-projects/jetbrains-ide-companion/build.gradle.kts:Project/TopLevel/stage2(local)})
        - InstrumentingVisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:]:root-project[:](export)})
        - VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:settings[:]:buildSrc[:]:kotlin-dsl:/media/disk/Documents/personal-projects/jetbrains-ide-companion/build.gradle.kts:Project/TopLevel/stage1(local)})
        - VisitableURLClassLoader(ClassLoaderScopeIdentifier.Id{coreAndPlugins:kotlin-dsl:/media/disk/Documents/personal-projects/jetbrains-ide-companion/settings.gradle.kts:Settings/TopLevel/stage1(local)})
Please report this error, run './gradlew --stop' and try again.
- Task `:ideaProject` of type `org.gradle.plugins.ide.idea.GenerateIdeaProject`: cannot serialize object of type 'org.gradle.api.internal.project.DefaultProject', a subtype of 'org.gradle.api.Project', as these are not supported with the configuration cache.
  See https://docs.gradle.org/8.14.3/userguide/configuration_cache.html#config_cache:requirements:disallowed_types

See the complete report at file:///media/disk/Documents/personal-projects/jetbrains-ide-companion/build/reports/configuration-cache/1xex43vs6m0e7eh1vx3kae308/dih55dlxx7y1b94jhcomrzhr0/configuration-cache-report.html
> Class 'org.gradle.plugins.ide.idea.IdeaPlugin$$Lambda/0x0000786108ad9a28' not found in Gradle runtime class loader 'MixInLegacyTypesClassLoader(legacy-mixin-loader)' of type 'org.gradle.initialization.MixInLegacyTypesClassLoader'.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 27s
4 actionable tasks: 4 executed
Configuration cache entry discarded with 6 problems.
```

<img width="1237" height="229" alt="Image" src="https://github.com/user-attachments/assets/09b52a1a-4b23-4d60-954d-20f475eb983b" />

--- a/src/main/kotlin/com/google/gemini/cli/IdeServer.kt
+++ b/src/main/kotlin/com/google/gemini/cli/IdeServer.kt
@@ -64,7 +64,7 @@ private data class PortInfo(
 
 class IdeServer(private val project: Project, private val diffManager: DiffManager) {
   private var server: HttpServer? = null
-  private var portFile: File? = null
+  private var portFiles: MutableList<File> = mutableListOf()
   private val authToken = generateAuthToken()
   private val openFilesManager: OpenFilesManager = project.service()
   private val transports = mutableMapOf<String, StreamableHttpServerTransport>()
@@ -373,14 +373,32 @@ class IdeServer(private val project: Project, private val diffManager: DiffManag
       )
     val portInfoJson = McpJson.encodeToString(portInfo)
 
+    // Clear old files
+    portFiles.forEach { it.delete() }
+    portFiles.clear()
+
+    // Discovery path: /tmp/gemini/ide/gemini-ide-server-${PID}-${PORT}.json
     val ideDir = File(System.getProperty("java.io.tmpdir"), "gemini/ide").apply { mkdirs() }
-    portFile = File(ideDir, "gemini-ide-server-$ppid-$actualPort.json").apply {
-      writeText(portInfoJson)
-      setReadable(true, true)
-      setWritable(true, true)
-      deleteOnExit()
+    
+    // We write files for the current PID and its parent to ensure
+    // gemini-cli can find the file regardless of how it traverses the process tree on Linux.
+    val pids = mutableListOf(ppid)
+    ProcessHandle.of(ppid).ifPresent { handle ->
+      handle.parent().ifPresent { p1 ->
+        pids.add(p1.pid())
+      }
     }
-    LOG.info("Wrote port info to ${portFile?.absolutePath}")
+
+    pids.distinct().forEach { pid ->
+      val file = File(ideDir, "gemini-ide-server-$pid-$actualPort.json")
+      file.writeText(portInfoJson)
+      file.setReadable(true, true)
+      file.setWritable(true, true)
+      file.deleteOnExit()
+      portFiles.add(file)
+    }
+
+    LOG.info("Wrote port info to ${portFiles.size} discovery files in ${ideDir.absolutePath}")
     project.service<GeminiCliServerState>().apply {
       this.port = actualPort
       this.token = authToken
@@ -445,7 +463,7 @@ class IdeServer(private val project: Project, private val diffManager: DiffManag
     transports.keys.toList().forEach { sessionId -> cleanupSession(sessionId) }
     coroutineScope.cancel()
     server?.stop(0)
-    portFile?.delete()
+    portFiles.forEach { it.delete() }
     project.service<GeminiCliServerState>().apply {
       this.port = null
       this.token = null

